src/Cache/MongoLock.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+namespace MongoDB\Laravel\Cache;
+
+use Illuminate\Cache\Lock;
+use MongoDB\Laravel\Collection;
+use MongoDB\Operation\FindOneAndUpdate;
+use Override;
+
+use function random_int;
+
+final class MongoLock extends Lock
+{
+    /**
+     * Create a new lock instance.
+     *
+     * @param  Collection  $collection              The MongoDB collection
+     * @param  string      $name
+     * @param  int         $seconds
+     * @param  string|null $owner
+     * @param  array       $lottery                 The prune probability odds
+     * @param int         $defaultTimeoutInSeconds The default number of seconds that a lock should be held
+     */
+    public function __construct(
+        private Collection $collection,
+        string $name,
+        int $seconds,
+        ?string $owner = null,
+        private array $lottery = [2, 100],
+        private int $defaultTimeoutInSeconds = 86400,
+    ) {
+        parent::__construct($name, $seconds, $owner);
+    }
+
+    /**
+     * Attempt to acquire the lock.
+     *
+     * @return bool
+     */
+    public function acquire()
+    {
+        // The lock can be acquired if: it doesn't exist, it has expired,
+        // or it is already owned by the same lock instance.
+        $condition = [
+            '$or' => [
+                ['$lte' => ['$expiration', $this->currentTime()]],
+                ['$eq' => ['$owner', $this->owner]],
+            ],
+        ];
+        $result = $this->collection->findOneAndUpdate(
+            ['key' => ['$eq' => $this->name]],
+            [
+                [
+                    '$set' => [
+                        'key' => [
+                            '$cond' => [
+                                'if' => $condition,
+                                'then' => $this->name,
+                                'else' => '$key',
+                            ],
+                        ],
+                        'owner' => [
+                            '$cond' => [
+                                'if' => $condition,
+                                'then' => $this->owner,
+                                'else' => '$owner',
+                            ],
+                        ],
+                        'expiration' => [
+                            '$cond' => [
+                                'if' => $condition,
+                                'then' => $this->expiresAt(),
+                                'else' => '$expiration',
+                            ],
+                        ],
+                    ],
+                ],
+            ],
+            [
+                'upsert' => true,
+                'returnDocument' => FindOneAndUpdate::RETURN_DOCUMENT_AFTER,
+            ],
+        );
+
+        if (random_int(1, $this->lottery[1]) <= $this->lottery[0]) {
+            $this->collection->deleteMany(['expiration' => ['$lte' => $this->currentTime()]]);
+        }
+
+        return $result->owner === $this->owner;
```

Comment:
Here I'm using `findOneAndUpdate` and a comparison of the value to return `true` when the lock is acquired twice by the same instance during the same second. If I use `updateOne` and check `$result->getUpsertedCount() > 0 || $result->getModifiedCount() > 0;`, the result is `false`.

Reply:
> when the lock is acquired twice by the same instance during the same second

What is the significance of "same instance during the same second"?

If `updateOne` reports that no document was upserted or modified, wouldn't that imply that the lock could not be acquired (i.e. has another owner and has not expired)?

I was testing this locally:

```php
$coll = $client->selectCollection('test', 'coll');
$coll->drop();

$key = 'foo';
$owner = 'bar';
$currentTime = time();
$expiresAt = $currentTime + 10;

$isExpiredOrAlreadyOwned = ['$or' => [
    ['$lte' => ['$expiration', $currentTime]],
    ['$eq' => ['$owner', $owner]],
]];

$result = $coll->updateOne(
    ['key' => $key],
    [
        ['$set' => [
            'owner' => ['$cond' => [$isExpiredOrAlreadyOwned, $owner, '$owner']],
            'expiration' => ['$cond' => [$isExpiredOrAlreadyOwned, $expiresAt, '$expiration']],
        ]],
    ],
    ['upsert' => true],
);

printf("\n\nupdateOne: matched(%d) modified(%d) upserted(%d)\n\n",
    $result->getMatchedCount(),
    $result->getModifiedCount(),
    $result->getUpsertedCount(),
);

if ($result->getMatchedCount()) {
    echo "lock '$key' already existed\n";
}

if ($result->getModifiedCount()) {
    echo "took over an expired lock for '$owner', which will now expire at $expiresAt\n";
}

if ($result->getUpsertedCount()) {
    echo "created the lock for '$owner', which will now expire at $expiresAt\n";
}
```

Reply:
I'm talking about this test:
https://github.com/mongodb/laravel-mongodb/blob/3fa18f3553e0907f8ac1d19d1560a84afddb9ff8/tests/Cache/MongoLockTest.php#L30-L34

`updateOne` reports no document modified if the values in `$set` are not different from the values already stored.

In your example, if you run `updateOne` 2 times, the 2nd time you get "lock 'foo' already existed" and nothing else. In Laravel, `true` is expected because the lock is acquired and owned by the current instance.

Reply:
I reverted to `updateOne` in this additional PR, because `UTCDateTime` is more precise and acquiring a lock with the same owner in the same microsecond will not happen.
https://github.com/mongodb/laravel-mongodb/pull/2891